dipy/reconst/ivim.py

```diff
@@ -0,0 +1,595 @@
+        data : array, (len(bvals))
+            An array containing the signal from a voxel.
+            If the data was a 3D image of 10x10x10 grid with 21 bvalues,
+            the multi_voxel decorator will run the single voxel fitting
+            on all the 1000 voxels to get the parameters in
+            IvimFit.model_paramters. The shape of the parameter array
+            will be (data[:-1], 4).
+
+        x0 : array
+            Initial guesses for the parameters S0, f, D_star and D
+            calculated using a linear fitting.
+
+        Returns
+        -------
+        x0 : array
+            Estimates of the parameters S0, f, D_star and D.
+        """
+        gtol = self.options["gtol"]
+        ftol = self.options["ftol"]
+        xtol = self.tol
+        epsfcn = self.options["eps"]
+        maxfev = self.options["maxiter"]
+        bounds = self.bounds
+
+        if SCIPY_LESS_0_17:
+            try:
+                res = leastsq(_ivim_error,
+                              x0,
+                              args=(self.gtab, data),
+                              gtol=gtol,
+                              xtol=xtol,
+                              ftol=ftol,
+                              epsfcn=epsfcn,
+                              maxfev=maxfev)
+                ivim_params = res[0]
+                return ivim_params
+            except:
+                return x0
+        else:
+            try:
+                res = least_squares(_ivim_error,
+                                    x0,
+                                    bounds=bounds,
+                                    ftol=ftol,
+                                    xtol=xtol,
+                                    gtol=gtol,
+                                    max_nfev=maxfev,
+                                    args=(self.gtab, data),
+                                    x_scale=self.x_scale)
+                ivim_params = res.x
+                return ivim_params
+            except:
+                return x0
+
+
+class IvimFit(object):
+    def __init__(self, model, model_params):
+        """ Initialize a IvimFit class instance.
+            Parameters
+            ----------
+            model : Model class
+            model_params : array
+                The parameters of the model. In this case it is an
+                array of ivim parameters. If the fitting is done
+                for multi_voxel data, the multi_voxel decorator will
+                run the fitting on all the voxels and model_params
+                will be an array of the dimensions (data[:-1], 4),
+                i.e., there will be 4 parameters for each of the voxels.
+        """
+        self.model = model
+        self.model_params = model_params
+
+    def __getitem__(self, index):
+        model_params = self.model_params
+        N = model_params.ndim
+        if type(index) is not tuple:
+            index = (index,)
+        elif len(index) >= model_params.ndim:
```

Comment:
These lines (L503-L505) are not currently being tested. Could you add a little test for error handling in indexing?